docs/RFCS/20220818_tenant_version_upgrades.md

```diff
@@ -0,0 +1,475 @@
+
+Once that cluster finishes upgrading its executable version so all
+nodes run at v20.2, it can *still* operate with logical version
+`v20.1-3344`, but now it can also be upgraded to `v20.2-1`.
+
+After the logical version has been upgraded to `v20.2-VVV`, it is not
+possible any more to downgrade the executable version to v20.1.
+
+## What gets controlled by logical versions
+
+We use logical versions as a control for two separate mechanisms:
+
+- logical versions are paired to *cluster upgrade migrations*: we run
+  changes to certain `system` tables and other low-level storage data
+  structures as a side effect of moving from one logical version to
+  another.
+
+- logical versions are also used as *feature gates*: during SQL
+  planning/execution, we check the current logical version and block
+  access to certain features as long as given logical version hasn't
+  been reached.
+
+The two are related: certain features require `system` tables /
+storage to be in a particular state. So we routinely introduce
+features with two logical versions: the first version upgrade
+"prepares" the system state, while the new feature is still
+inaccessible from SQL; then, the second version upgrade enables the
+SQL feature.
+
+## Additional invariants
+
+For the above mechanisms to work, we need the following invariants:
+
+- the logical version must evolve monotonically, that is it never goes
+  down. In particular, the SQL code must always *observe* it to
+  increase monotonically.
+
+  We need this because the `system` table / storage upgrades are not
+  reversible, and once certain SQL features are enabled we can't
+  cleanly un-enable them (e.g. temp tables).
+
+- all nodes in a cluster either see logical version X or X-1. There's
+  never a gap of more than 1 version across the cluster.
+
+  We need this because we also remove features over time. It's
+  possible for version X+1 to introduce a replacement feature, then
+  version X+2 to remove the old feature. If we allowed versions X and
+  X+2 to be visible at the same time, the feature could be both
+  enabled and disabled (or both old and new) in different parts of the
+  cluster at the same time. We don't want that.
+
+## Where the logical version is persisted
+
+The logical version is not a regular cluster setting. It is persisted
+in two difference places:
+
+- it is persisted in the `system.settings` table, for the benefit of
+  SHOW CLUSTER SETTINGS and other "superficial" SQL observability
+  features.
+
+- more importantly, it is stored in a reserved configuration field
+  on each store (i.e., per store on each KV node).
+
+The invariants are then maintained during the *upgrade and migration*
+process, described below.
+
+The special case of cluster creation (when there is no obvious "previous
+version") is also handled via the upgrade process, as follows:
+
+- When a new cluster gets initially created, the initial nodes write
```

Comment:
I was tracing this recently and it looked like they start by writing their _min_ binary version and then upgrade it to binary version, don't they?